### ISSUE TITLE
invoke method is not working

### ISSUE THREAD

**Issue by IanTrudel** [uniconproject/unicon#422]
The package lang provides the utility method invoke but it does not work. It was confirmed when I ran objecttests. Unicon version 13.3 on Linux.

Output from unicon/tests/lib/objecttests.icn:

```
Check to see if d is an instance of Class.

d is NOT an instance of Class!

Test string invocation of method names...should see '123'


List all the methods in class D...

DTest_write

Check the inheritance hierarchy

ATest is NOT an instance of Class
ATest is NOT an instance of BTest
BTest is NOT an instance of ATest
BTest is NOT an instance of BTest
CTest is NOT an instance of Class
CTest is NOT an instance of ATest
DTest is NOT an instance of Class
DTest is NOT an instance of ATest
DTest is NOT an instance of BTest
DTest is NOT an instance of CTest
ETest is NOT an instance of Class
ETest is NOT an instance of FTest
FTest is NOT an instance of ETest

d inherits from: DTest ATest CTest lang::Object

e inherits from: ETest FTest lang::Object ETest

g fields: one two three
h fields: four five six one two three
h.one -> 1
h.one -> One
Attempt to set non-existent field: ok - field doesn't exist
```

**Comment by StephenWampler** [comment]
Okay, there are lots of interesting things going wrong here.  Starting with the simple and heading towards the complex:

1) The class "Object" no longer subclasses the class "Class", so objecttests.icn needs to change all references to the Class class into references to the Object class.

2) The procedure instanceof in uni/lib/predicat.icn currently calls the istype procedure.  istype appears to be broken, possibly in several ways (for one thing it calls get_type but get_type doesn't traverse the inheritance hierarchy).  A workaround would be to modify the instanceof method to call lang::Type instead of istype.  The correct thing may be to fix get_type but who knows what that may break?

3) In uni/lib/object.icn, the method invoke calls hasMethod, which has the line:

    return mName == genMethods()

however, genMethods now calls the builtin function methodnames to generate the names of methods.  But methodnames returns the *internal* names of class methods, e.g. "DTest__write".  I think this is wrong but, again, what breaks if it's fixed?   One would think that a workaround would be to modify genmethods to map the internal names back into external ones.  That's not sufficient because invoke(), assuming that's fixed, will die because the table lookup it uses (self.__m[mName]) will fail because the external name, e.g. "DTest::write" won't be found in __m - it needs to be just "write", so a 2nd workaround is needed to map the full external method name into its simple form.  Note that fixing the 'wrong' methodnames function to return the simple form of method names would be sufficient unto itself (he claims).

**Comment by IanTrudel** [comment]
This is an amazingly insightful reply. Thank you! My current workaround is to use `__m[mname]` but it's not without its flaws. It requires the receiver object to be passed on as the first parameter, i.e. `object.__m[nmane](object)`, otherwise it will send it to self. In the following code example, you will notice that the output is `Child::print()` instead of `Parent::print()`, which is not a typical behaviour for OOP. It would most likely require to call `o.__m[mname](self)` in the run method of the Parent class. 

Output:
```
Anonymous::print()
Child::print()
```

```
class Parent(__objects)
   method add(o)
      put(__objects, o)
   end

   method print()
      write(classname(self) || "::print()")
   end

   method run()
      local mname

      every o := !__objects do {
         cname := classname(o)

         every m := !methodnames(o) do {
            m ? {
               tab(upto(cname || "_")) & move(*cname + 1)
               mname := tab(0)
            }
            o.__m[mname](o)
         }
      }
   end

initially
   __objects := []
end

class Child : Parent()
   method print()
      self.Parent.print()
   end
end

class Anonymous()
   method print()
      write(classname(self) || "::print()")
   end
end

procedure main()
   p := Parent()
   a := Anonymous()
   c := Child()
   p.add(a)
   p.add(c)
   p.run()
end
```

**Comment by StephenWampler** [comment]
I'm not sure what's causing the "Child::print()" output or even it's the "Right Thing To Do" but one suggestion for your code: in place of upto(cname||"_"), use find(cname||"_")  the upto function converts its first argument to a cset and goes up to the first appearance of any character in that cset.  You're actually lucky that the class names you've got work with upto.

I assume also that this is a test program to explore behavior.  Ordinarily, omitting the print method from Child would default a call to print to the Parent's print()  [but still output "Child::print()", of course].  However, the run() method won't find the Parent's print method in that case.  I suspect you know that and are just experimenting.

**Comment by StephenWampler** [comment]
Ah, the reason run() doesn't find the print() method in Parent when the subclass doesn't have it is because methodnames() is only returning the 'local' methods, not any derived from ancestors.  This is, he claims, another problem with the methodnames implementation.  The __m list has all the methods (included inherited ones) in it but methodnames() only provides the local ones.   I ;think fixing methodnames() and invoke() may be non-trivial.

**Comment by StephenWampler** [comment]
I played around a bit with your test program and made some changes.  The big ones are that the run() method:

(a)  now takes as a argument the name of the method (e.g. "print") to run.  This version doesn't accept any arguments to pass to that method - that should be fixed.

(b) now can find a method that exists back in the class hierarchy (it avoids using the broken methodnames)

(c)  only involkes the first instance of a method from the hierarchy, starting from the 'local' and going up into superclasses

Assuming the issue stated in (a) is addressed, I think the run() method points to a viable replacement for the invoke() method.

Anyway, the amended code:

class Parent(__objects)
   method add(o)
      put(__objects, o)
   end

   method print()
      write(classname(self) || "::print()")
   end

   method run(s)   # Pass in name of method to run, should also pass in args but doesn't yet...
      local mname

      every o := !__objects do {
         cname := classname(o)

         # If we find that method, then call it and stop looking for it!
         #   Note: this is prelim version that doesn't pass any args to method.
         every m := image(!o.__m) do {
            m ? {
               mname := (="procedure ", tab(0))  | next
               if mname ? (tab(-(*s+1)),="_",match(s)) then break mname(o)
            }
         }
      }
   end

initially
   __objects := []
end

class Child : Parent()
end

class Anonymous()
   method print()
      write(classname(self) || "::print()")
   end
end

procedure main()
   p := Parent()
   a := Anonymous()
   c := Child()
   p.add(a)
   p.add(c)
   p.run("print")
end

**Comment by brucerennie** [comment]
Good morning Ian,

For my purposes, I built a class hierarchy to get around some of the 
problems within the library Class/Object that I found. My particular 
solution is found in the two files in lib called classobject.icn and 
errorsystem.icn.

This is used in utf8.icn (also in lib).

Fundamentally, the problem is that there is no direct correspondence 
between the user visible package/class/method names and the internal 
names. I am looking at a possible change here, but only after I finish 
the new keyword additions.

regards and blessings

Bruce Rennie

**Comment by StephenWampler** [comment]
Oops.  Typo.  The change would be to uni/lib/object.icn, not uni/lib/class.icn.  Sorry!

**Comment by Jafaral** [comment]
> Oops. Typo. The change would be to uni/lib/object.icn, not uni/lib/class.icn. Sorry!

You can edit a previous post by pressing the three dots at the top right corner of post.

**Comment by Jafaral** [comment]
@IanTrudel tests are very welcome, thank you for doing the work.

FYI, we do have tests run as part of the CI  ([sample](https://github.com/uniconproject/unicon/actions/runs/8733300130/job/23961740833)). We are  close to changing that so that any failed test would fail the CI. We are in the process in fixing the remaining failing tests or at least isolating them.

**Comment by IanTrudel** [comment]
> @IanTrudel tests are very welcome, thank you for doing the work.

My pleasure! `tests/unicon/tester.icn` is very good but it's rather minimalist. With a bit of guidance from you guys, I can come up with a framework that offers convenience, high cohesion and extensibility.

> FYI, we do have tests run as part of the CI (see Test [here](https://github.com/uniconproject/unicon/actions/runs/8733300130/job/23961740833)). We are close to changing that so that any failed test would fail the CI. We are in the process of fixing the remaining failing tests or at least isolating them.

Great work on the CI, by the way! I was surprised to see support for so many targets.

**Comment by IanTrudel** [comment]
> Here's a potential replacement for the invoke() method in uni/lib/object.icn. It avoids the problem of mapping between internal/external/simple method names. I've only tested it a little, so caveat emptor!
> 
> method invoke(s, A[]) local mname
> 
> ```
>   every image(!self.__m) ? {
>      mname := (="procedure ", tab(0)) | next
>      if mname ? (tab(-(*s+1),="_",match(s)) then
>            suspend mname ! ([self]|||A) | fail
>      }
> ```
> 
> end

@StephenWampler Your solution turned out pretty well. It's clean and straightforward.

```
      every test := !__tests do {
         cname := classname(test)

         test.setupClass()
         every image(!test.__m) ? {
            mname := (="procedure ", tab(0)) | next
            if mname ? (tab((*cname+1)), ="_", match("test")) then {
               test.setup()
               mname(test)
               test.teardown()
            }
         }
         test.teardownClass()
      }
```

**Comment by Jafaral** [comment]
@StephenWampler, wanna go after some of the issues you identified? We should attempt to fix those. If they break any existing applications we can discuss.

**Comment by StephenWampler** [comment]
Invoke should be fixed with issue #441.  Also, despite multiple inheritance, only one of any duplicated method is ever visible to the subclass.  There's a fix in #441 so all of the visible methods available to a subclass are produced by genMethods().  objecttests.icn now produces the expected output.  Note that the function (i.e. "builtin") methodnames is still broken.  My C is so far in the past the I'm absolutely the wrong person to try and fix it.  I know.  I looked.  It's pretty much greek to me...

**Comment by StephenWampler** [comment]
I'm going to close this issue since invoke() now works as expected.  I'm also going to open a new issue on the fact that methodnames() doesn't generate inherited methods visible to the subclass.
